### PR TITLE
Feature branch: bulk attribute groups

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

**Review by StevenVAIDIE — COMMENTED** [review]
GTM

**Review by ValentinMumble — APPROVED** [review]

File: src/Akeneo/Pim/Structure/Bundle/Domain/Query/Sql/GetAttributeGroupsInterface.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * @copyright 2023 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+namespace Akeneo\Pim\Structure\Bundle\Domain\Query\Sql;
+
+interface GetAttributeGroupsInterface
+{
+    public function all(): array;
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Domain/Query/Sql/GetAttributeGroupsInterface.php:14]: This is not a good idea IMHO.
We don't have any limits for attribute groups, nothing forbids user to create 10k attribute groups.
2 solutions:
- add a hard limit on attribute group count, change doc, limit has to be decided with popeye/pautophe
- update this method to return an iterator. The internal function should get items with paginated queries.

File: src/Akeneo/Pim/Structure/Bundle/Query/InternalApi/AttributeGroup/Sql/GetAttributeGroups.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+GROUP BY code, sort_order
+ORDER BY sort_order;
+SQL;
+
+        $attributeGroups = $this->connection->executeQuery($sql)->fetchAllAssociative();
+
+        return array_map(function (array $attributeGroup) {
+            $attributeGroup['sort_order'] = (int) $attributeGroup['sort_order'];
+            $attributeGroup['labels'] = json_decode($attributeGroup['labels'], true);
+            // TODO: check FT and call ServiceAPI
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Query/InternalApi/AttributeGroup/Sql/GetAttributeGroups.php:37]: don't let TODO or add the ticket number next to it :)

File: src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml [hunk 3]
```diff
@@ -273,8 +280,33 @@ pim_enrich.entity.attribute_group:
             confirm: Are you sure you want to delete this attribute group?
     grid:
         columns:
-            name: Attribute groups name
-        no_search_result: Sorry, there is no result for your search.
+            attribute_count: Related attributes count
+    selected: '{0}0 attribute group selected|{1}1 attribute group selected|]1, Inf[{{ count }} attribute groups selected'
+    mass_delete:
+        button: Delete
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml:286]: this already exists in the pim_common

File: src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml [hunk 4]
```diff
@@ -273,8 +280,33 @@ pim_enrich.entity.attribute_group:
...
+        title: Attribute Groups
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml:287]: pim_enrich.entity.attribute_group.plural_label

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx [hunk 5]
```diff
@@ -1,47 +1,82 @@
...
+import {useAttributeGroupPermissions, useAttributeGroupsIndexState, useFilteredAttributeGroups} from '../../../hooks';
 
-const FeatureFlags = require('pim/feature-flags');
+const Wrapper = styled.div`
+  display: flex;
+  flex-direction: column;
+`;
+
+const TableWrapper = styled.div`
+  // TODO uncomment when bringing back selectable margin-left: -40px;
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx:15]: samesame

File: src/Akeneo/Pim/Structure/Bundle/Domain/Query/Sql/GetAttributeGroupsInterface.php (hunk 1 above)
Comment (MadeWilson) [reply src/Akeneo/Pim/Structure/Bundle/Domain/Query/Sql/GetAttributeGroupsInterface.php:14]: We will add a hard limit to 1000 https://akeneo.atlassian.net/browse/RAB-1300

File: src/Akeneo/Pim/Structure/Bundle/Query/InternalApi/AttributeGroup/Sql/GetAttributeGroups.php (hunk 2 above)
Comment (MadeWilson) [reply src/Akeneo/Pim/Structure/Bundle/Query/InternalApi/AttributeGroup/Sql/GetAttributeGroups.php]: dead code removed

File: src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml (hunk 3 above)
Comment (MadeWilson) [reply src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml]: replaced

File: src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml (hunk 4 above)
Comment (MadeWilson) [reply src/Akeneo/Pim/Structure/Bundle/Resources/translations/jsmessages.en_US.yml]: replaced

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx (hunk 5 above)
Comment (MadeWilson) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx]: ticket number added

File: src/Akeneo/Pim/Structure/Bundle/EventSubscriber/CheckAttributeGroupOtherCannotBeRemovedSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public function preRemove(RemoveEvent $event): void
+    {
+        $attributeGroup = $event->getSubject();
+        if (!$attributeGroup instanceof AttributeGroupInterface) {
+            return;
+        }
+
+        if ('other' === $attributeGroup->getCode()) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/EventSubscriber/CheckAttributeGroupOtherCannotBeRemovedSubscriber.php:34]: you could use the constant here (bonus points if you move it in AttributeGroupInterface)

File: src/Akeneo/Pim/Structure/Bundle/Infrastructure/Controller/GetAttributeGroupController.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public function __construct(
+        private readonly SecurityFacadeInterface $securityFacade,
+        private readonly GetAttributeGroupHandler $getAttributeGroupHandler
+    ) {
+    }
+
+    public function __invoke(): Response
+    {
+        if (!$this->securityFacade->isGranted('pim_api_attribute_group_list')) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Structure/Bundle/Infrastructure/Controller/GetAttributeGroupController.php:28]: wrong ACL

**Review by mmetayer — APPROVED** [review]
GTM once ACL is fixed

**Review by guillaumefer-akeneo — APPROVED** [review]

**Pull request merged**